res/controllers/Pioneer-DDJ-SR2-scripts.js

```diff
@@ -0,0 +1,1600 @@
+DDJSR2.rotarySelectorChanged = false;
+DDJSR2.panels = [false, false]; // view state of effect and sampler panel
+// DDJSR2.shiftPanelSelectPressed = false;
+
+DDJSR2.syncRate = [0, 0, 0, 0];
+DDJSR2.needleSearchTouched = [false, false, false, false];
+DDJSR2.chFaderStart = [null, null, null, null];
+DDJSR2.toggledBrake = [false, false, false, false];
+DDJSR2.scratchMode = [true, true, true, true];
+DDJSR2.wheelLedsBlinkStatus = [0, 0, 0, 0];
+DDJSR2.setUpSpeedSliderRange = [0.08, 0.08, 0.08, 0.08];
+DDJSR2.serato=[0xF0, 0x00, 0x20, 0x7F, 0x50, 0x01, 0xF7];
+DDJSR2.seratoPoll=[0xF0, 0x00, 0x20, 0x7F, 0x03, 0x01, 0xF7];
+DDJSR2.selectedSamplerBank = 0;
+DDJSR2.fxKnobMSBValue = [0, 0];
+DDJSR2.shiftFxKnobMSBValue = [0, 0];
+
+
+DDJSR2.valueVuMeter = {
+    "[Channel1]_current": 0,
+    "[Channel2]_current": 0,
+    "[Channel3]_current": 0,
+    "[Channel4]_current": 0,
+    "[Channel1]_enabled": 1,
+    "[Channel2]_enabled": 1,
+    "[Channel3]_enabled": 1,
+    "[Channel4]_enabled": 1
+};
+
+DDJSR2.channelGroups = {
+    "[Channel1]": 0x00,
+    "[Channel2]": 0x01,
+    "[Channel3]": 0x02,
+    "[Channel4]": 0x03
+};
+
+DDJSR2.scratchSettings = {
+    "alpha": 1.0 / 8,
+    "beta": 1.0 / 8 / 32,
+    "jogResolution": 800,
+    "vinylSpeed": 33 + 1 / 3,
+};
+
+DDJSR2.wheelLedCircle = {
+    "minVal": 0,
+    "maxVal": 0x14,
+    "blink": 0x3
+};
+
+DDJSR2.fx1assignLeds = [0x4C, 0x50, 0x70, 0x54];
+DDJSR2.fx2assignLeds = [0x4D, 0x51, 0x71, 0x55];
+
+////////////
+// Code. //
+//////////
+
+DDJSR2.init = function() {
+    let i, j;
+
+
+    DDJSR2.deck = [];
+    for (i = 0; i < 4; i++) {
+        DDJSR2.deck[i] = new DDJSR2.Deck(i + 1, i);
+        DDJSR2.deck[i].setCurrentDeck("[Channel" + (i + 1) + "]");
+    }
+
+    DDJSR2.effectUnit = [];
+    for (i = 1; i <= 2; i++) {
+        DDJSR2.effectUnit[i] = new components.EffectUnit([i, i + 2]);
+        DDJSR2.effectUnit[i].enableButtons[1].midi = [0x93 + i, 0x47];
+        DDJSR2.effectUnit[i].enableButtons[2].midi = [0x93 + i, 0x48];
+        DDJSR2.effectUnit[i].enableButtons[3].midi = [0x93 + i, 0x49];
+        DDJSR2.effectUnit[i].dryWetKnob.input = function(channel, control, value, _status, _group) {
+            this.inSetParameter(this.inGetParameter() + DDJSR2.getRotaryDelta(value) / 30);
+        };
+        for (j = 1; j <= 4; j++) {
+            DDJSR2.effectUnit[i].enableOnChannelButtons.addButton(`Channel${  j}`);
+            var midiout;
+            if (i === 1) {
+                midiout = DDJSR2.fx1assignLeds[j-1];
+            } else {
+                midiout = DDJSR2.fx2assignLeds[j-1];
+            }
+            DDJSR2.effectUnit[i].enableOnChannelButtons[`Channel${  j}`].midi = [0x96, midiout];
+        }
+
+        DDJSR2.effectUnit[i].init();
+    }
+
+    // Get initial state of controls:
+    midi.sendShortMsg(0x9B, 0x08, 0x7F);
+
+    DDJSR2.seratoTimer = engine.beginTimer(250, "DDJSR2.doTimer", 0);
+    midi.sendSysexMsg(DDJSR2.seratoPoll, DDJSR2.seratoPoll.length);
```

Comment:
I would assume from the comment, that it instructs the SR2 to dump the current control state. Will confirm once I've changed the rest of the mapping, according to the review.

Reply:
Please do. I doubt the midi message does that, usually the sysex message does that.

Reply:
I think that the seratopoll method might have to do with unlocking the controller, since this model is a serato one, not a generic midi one.
On the hercules dj console 4-mx, there is also a midi message to tell the controller to send the control values, and they are received as if they had just been tweaked.
I am assuming this controller supports that too.